arwen/host/execution_async_context.go

```diff
@@ -0,0 +1,435 @@
+		}
+
+		if asyncCallGroup.IsCompleted() {
+			asyncContext.DeleteAsyncCallGroup(groupID)
+		}
+	}
+
+	// Step 2: redistribute unspent gas; then, in one combined step, do the
+	// following:
+	// * locally execute built-in functions with cross-shard
+	//   destinations, whereby the cross-shard OutputAccount entries are generated
+	// * call host.sendAsyncCallCrossShard() for each pending AsyncCall, to
+	//   generate the corresponding cross-shard OutputAccount entries
+	host.setupAsyncCallsGas(asyncContext)
+
+	for _, asyncCallGroup := range asyncContext.AsyncCallGroups {
+		// Execute the call group allowing asynchronous (cross-shard) calls as well
+		err := host.executeAsyncCallGroup(asyncCallGroup, false)
+		if err != nil {
+			return err
+		}
+	}
+
+	host.saveAsyncContext(asyncContext)
+
+	return nil
+}
+
+func (host *vmHost) executeAsyncCallGroup(
+	asyncCallGroup *arwen.AsyncCallGroup,
+	syncExecutionOnly bool,
+) error {
+	for _, asyncCall := range asyncCallGroup.AsyncCalls {
+		err := host.executeAsyncCall(asyncCall, syncExecutionOnly)
+		if err != nil {
+			return err
+		}
+	}
+
+	asyncCallGroup.DeleteCompletedAsyncCalls()
+
+	// If ALL the AsyncCalls in the AsyncCallGroup were executed synchronously,
+	// then the AsyncCallGroup can have its callback executed.
+	if asyncCallGroup.IsCompleted() {
+		// TODO reenable this, after allowing a gas limit for it and deciding what
+		// arguments it receives (this method is currently a NOP and returns nil)
+		return host.executeAsyncCallGroupCallback(asyncCallGroup)
+	}
+
+	return nil
+}
+
+func (host *vmHost) executeAsyncCall(
+	asyncCall *arwen.AsyncCall,
+	syncExecutionOnly bool,
+) error {
+	execMode, err := host.determineAsyncCallExecutionMode(asyncCall)
+	if err != nil {
+		return err
+	}
+
+	if execMode == arwen.SyncExecution {
+		vmOutput, err := host.executeSyncCall(asyncCall)
+		// TODO host.executeSyncCallback() returns a vmOutput produced by executing
+		// the callback. Information from this vmOutput should be preserved in the
+		// pending AsyncCallGroup, and made available to the callback of the
+		// AsyncCallGroup (currently not implemented).
+		// Either way, host.executeSyncCallback() calls
+		// host.ExecuteOnDestContext(), which merges the vmOutput of the callback
+		// into the main output (such as OutputAccounts).
+		callbackVMOutput, callbackErr := host.executeSyncCallback(asyncCall, vmOutput, err)
+		host.finishSyncExecution(callbackVMOutput, callbackErr)
+		return nil
+	}
+
+	if syncExecutionOnly {
+		return nil
+	}
+
+	if execMode == arwen.AsyncBuiltinFunc {
+		// Built-in functions will handle cross-shard calls themselves, by
+		// generating entries in vmOutput.OutputAccounts. It is not necessary to
+		// call sendAsyncCallCrossShard(). The vmOutput produced by the built-in
+		// function, containing the cross-shard call, has ALREADY been merged into
+		// the main output by the inner call to host.ExecuteOnDestContext().
+		_, err := host.executeSyncCall(asyncCall)
```

Comment:
shouldn't we treat returnCode != vmcommon.OK from builtInFuncCall ?

Reply:
Good catch. Added sync callback execution in case the return code if it's not `vmcommon.Ok`, so that the contract can handle the early failure, even if the cross-shard call has not yet left the VM.

And if the built-in function returns `vmcommon.Ok`, no sync callback will be executed. Instead, the corresponding callback will be called *when returning* from the cross-shard call.

Reply:
Failed built in func call will not leave the current shard. It is only executed at sender. Only successful ones are sent to the other shard.

Reply:
Yes, the new condition takes that into account.